corehq/apps/users/tasks.py

```diff
@@ -36,6 +38,21 @@ def tag_docs_as_deleted(cls, docs, deletion_id):
     cls.get_db().bulk_save(docs)
 
 
+@task(rate_limit=2, queue='background_queue', ignore_result=True)  # 2 saves/sec for cloudant slowness
+def tag_forms_as_deleted_rebuild_associated_cases(formlist, deletion_id, deleted_cases=set()):
+    from casexml.apps.case.cleanup import rebuild_case
+
+    cases_to_rebuild = set()
+    for form in formlist:
+        form['doc_type'] += DELETED_SUFFIX
+        form['-deletion_id'] = deletion_id
+        for case in get_case_ids_from_form(form) - deleted_cases:
+            cases_to_rebuild.add(case)
```

Comment:
Sorry to be picky but if `get_case_ids_from_form` returns a set you could just do a union / update operation to add them all to `cases_to_rebuild`. You could also do the `- deleted_cases` at the end instead of inside the loop.